### PR TITLE
(SERVER-876) Add code_id to catalog requests

### PR DESCRIPTION
**Pull request by haus**
This commit leverages the code-id-command in the versioned-code-service
to calculate and add the code-id to the catalog request. This is done by
adding a special handler for catalog requests that first gets the
environment out of the request and then uses that to calculate the code
id.

### CODE REVIEW

**Comment by haus** [comment]
Blocked by https://github.com/puppetlabs/puppet-server/pull/865

**Comment by haus** [comment]
Just updated to include the commits in #865 .

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 1]
```diff
@@ -37,9 +39,11 @@
                    (request-handler request))
 
     (comidi/GET ["/catalog/" [#".*" :rest]] request
-                   (request-handler request))
+                   ((request-core/catalog-request-handler
+                     request-handler get-code-id-fn) request))
     (comidi/POST ["/catalog/" [#".*" :rest]] request
-                    (request-handler request))
+                    ((request-core/catalog-request-handler
+                      request-handler get-code-id-fn) request))
```
Comment (KevinCorcoran) [inline src/clj/puppetlabs/services/master/master_core.clj]: This is a bit hard to parse, what with the invocation of a function that returns a function which is then invoked.  Perhaps something like that
```clj
(let [request-handler-with-code-id (request-core/catalog-request-handler request-handler get-code-id-fn)]
  (request-handler-with-code-id request)
```
... or something like that.

File: src/clj/puppetlabs/services/request_handler/request_handler_core.clj [hunk 2]
```diff
@@ -325,6 +325,20 @@
       (.handleRequest (:jruby-instance request))
       response->map)))
 
+(defn get-environment-from-request
+  "Gets the environment from a request, after the params map has been populated
+  via wrap-params-for-jruby."
+  [req]
+  ((:params (wrap-params-for-jruby req)) "environment"))
```
Comment (KevinCorcoran) [inline src/clj/puppetlabs/services/request_handler/request_handler_core.clj]: How's about:
```clj
(-> req
    wrap-params-for-jruby
    (get-in [:params "environment"]))

File: src/clj/puppetlabs/services/request_handler/request_handler_core.clj [hunk 3]
```diff
@@ -325,6 +325,20 @@
...
+
+(defn catalog-request-handler
+  "Specialized handler for catalogs. Adds code_id to the request before passing
```
Comment (KevinCorcoran) [inline src/clj/puppetlabs/services/request_handler/request_handler_core.clj]: This is actually a middleware, not a handler.

**Comment by haus** [comment]
I'm not entirely happy with where the integration test ended up. I may move it to the master_core test, although that lives in the unit directory.

File: src/clj/puppetlabs/services/request_handler/request_handler_core.clj (hunk 3 above)
Comment (haus) [inline src/clj/puppetlabs/services/request_handler/request_handler_core.clj]: catalog-request-middleware then?

**Comment by KevinCorcoran** [comment]
I'd worry more about getting the test in the correct namespace and not get hung up on the `unit` and `integration` directories.  We've talked about getting rid of those in the past, the idea being that a single namespace almost always has both unit and integration tests in it.

File: src/clj/puppetlabs/services/request_handler/request_handler_core.clj (hunk 3 above)
Comment (KevinCorcoran) [inline src/clj/puppetlabs/services/request_handler/request_handler_core.clj]: The usual naming convention for middleware is `wrap-*` or `wrap-with-*`.   So I'd actually go `wrap-with-code-id` or something like that.

File: test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj [hunk 4]
```diff
@@ -38,3 +37,27 @@
...
+(deftest ^:integration code-id-request-test
+  (testing "code id is added to the request body for catalog requests"
+    ; As we have set code-id-command to echo, the code id will
+    ; be the result of running `echo $environment`, which will
+    ; be production here.
+    (bootstrap/with-puppetserver-running
+     app {:jruby-puppet
+          {:max-active-instances num-jrubies}
+          :versioned-code
+          {:code-id-command (script-path "echo")}}
```
Comment (KevinCorcoran) [inline test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj]: Just looked at `script-path` and it refers to `shell_utils_test`.  That doesn't have anything to do with this test, so ... probably worth either changing the definition of `script-path` or using something else here.

File: test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj [hunk 5]
```diff
@@ -38,3 +37,27 @@
...
+    ; As we have set code-id-command to warn, the code id will
+    ; be the result of running `warn $environment`, which will
+    ; exit non-zero and return nil.
+    (bootstrap/with-puppetserver-running
+     app {:jruby-puppet
+          {:max-active-instances num-jrubies}
+          :versioned-code
+          {:code-id-command (script-path "warn")}}
+     (let [catalog (get-catalog)]
+       (is nil? (get catalog "code_id"))))))
```
Comment (KevinCorcoran) [inline test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj]: perhaps this should also test that a message was logged with the output from STDOUT and/or STDERR

File: test/unit/puppetlabs/services/master/master_core_test.clj [hunk 6]
```diff
@@ -63,6 +65,35 @@
           (is (= "something-crazy/for-content-type"
                  (get-in resp [:request :content-type]))))))))
 
+(deftest code-id-injection-test
+  (testing "code_id is calculated and added to the catalog request."
+    (let [handler (fn ([req] {:request req}))
+          app     (build-ring-handler handler "1.2.3" (constantly "foo-is-my-codeid"))
+          resp    (app {:request-method :get
+                        :uri            "/v3/catalog/bar"})]
```
Comment (KevinCorcoran) [inline test/unit/puppetlabs/services/master/master_core_test.clj]: Our style is to not align these things vertically.   https://github.com/puppetlabs/pl-clojure-style

File: test/unit/puppetlabs/services/master/master_core_test.clj [hunk 7]
```diff
@@ -63,6 +65,35 @@
...
+      (is (= "foo-is-my-codeid" (get-in resp [:request :params "code_id"])))))
+  (testing "code_id is not added to non-catalog requests"
+    (let [handler (fn ([req] {:request req}))
+          app (build-ring-handler handler "1.2.3" (constantly "foo-is-my-codeid"))
+          request (fn r ([path] (r :get path))
+                    ([method path] (app (mock/request method path))))]
```
Comment (KevinCorcoran) [inline test/unit/puppetlabs/services/master/master_core_test.clj]: I think `r` is unnecessary here.

It's also really weird to see an anonymous function with multiple arities ... don't think I've ever seen that.  Maybe that's why you named it `r`?  I'd probably just pull this out to a top-level function.

Comment (haus) [inline test/unit/puppetlabs/services/master/master_core_test.clj]: that is all copy-pasta from above (line 25), so pulling it out into a top level function makes sense to me.

File: test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj (hunk 5 above)
Comment (haus) [inline test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj]: sounds reasonable.

File: test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj (hunk 4 above)
Comment (haus) [inline test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj]: I think I'll be adding a commit to rearrange some common shared test bits, so this will get cleaned up in that refactor.

Comment (haus) [inline test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj]: I moved all of the shell scripts into `dev-resources/puppetlabs/common` and updated the tests to match.

**Comment by haus** [comment]
@KevinCorcoran I think I've addressed all of your feedback.

File: test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj (hunk 5 above)
Comment (haus) [inline test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj]: Something about testing this isn't working. the matcher thinks nothing is being logged.

File: test/unit/puppetlabs/services/master/master_core_test.clj [hunk 8]
```diff
@@ -63,6 +68,34 @@
...
+  (testing "code_id is calculated and added to the catalog request."
+    (let [handler (fn ([req] {:request req}))
+          app     (build-ring-handler handler "1.2.3" (constantly "foo-is-my-codeid"))
+          resp    (app {:request-method :get
+                        :uri "/v3/catalog/bar"})]
+      (is (= "foo-is-my-codeid" (get-in resp [:request :params "code_id"])))))
+  (testing "code_id is not added to non-catalog requests"
+    (let [handler (fn ([req] {:request req}))
+          app (build-ring-handler handler "1.2.3" (constantly "foo-is-my-codeid"))
+          request     (partial app-request app)]
```
Comment (jpinsonault) [inline test/unit/puppetlabs/services/master/master_core_test.clj]: Not important, but the let alignment on this and the `testing` block above are inconsistent. I think we're not vertically aligning them?

**Comment by jpinsonault** [comment]
I'm :+1: looks good

File: test/unit/puppetlabs/services/master/master_core_test.clj (hunk 8 above)
Comment (haus) [inline test/unit/puppetlabs/services/master/master_core_test.clj]: fixed i think?

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 9]
```diff
@@ -19,39 +20,41 @@
...
+                 (request-handler request))
+     (comidi/GET ["/file_metadata/" [#".*" :rest]] request
+                 (request-handler request))
+     (comidi/GET ["/file_bucket_file/" [#".*" :rest]] request
+                 (request-handler request))
+     (comidi/PUT ["/file_bucket_file/" [#".*" :rest]] request
+                 (request-handler request))
+     (comidi/HEAD ["/file_bucket_file/" [#".*" :rest]] request
+                  (request-handler request))
+
```
Comment (KevinCorcoran) [inline src/clj/puppetlabs/services/master/master_core.clj]: this newline bugs me.  perhaps a little too much.  :bug:

Comment (KevinCorcoran) [inline src/clj/puppetlabs/services/master/master_core.clj]: and now I realize that you did not add it.

File: test/unit/puppetlabs/services/master/master_core_test.clj [hunk 10]
```diff
@@ -63,6 +68,34 @@
...
+                     "file_bucket_file"
+                     "resource_type"
+                     "resource_types"
+                     "status"]
+               :put ["file_bucket_file"
+                     "report"]
+               :head ["file_bucket_file"]}
+              path paths]
+        (let [resp (request method (str "/v3/" path "/bar"))]
+          (is (= "wtf-no-code-id" (get-in resp [:request :params "code_id"] "wtf-no-code-id"))))))))
```
Comment (KevinCorcoran) [inline test/unit/puppetlabs/services/master/master_core_test.clj]: This assertion would be clearer without the default value, so just `(is (nil? (get-in resp ...`

**Comment by haus** [comment]
updated
